### PR TITLE
Implement Arbitrary for &[u8].

### PR DESCRIPTION
**Pull request by frewsxcv**
https://github.com/rust-fuzz/arbitrary/issues/62#issuecomment-736041629

### CODE REVIEW

File: src/lib.rs
Comment (frewsxcv) [inline src/lib.rs:636]: I cargo culted this line from the `Vec<A>` implementation. Can someone explain why we defer to the `usize` `Arbitrary` implementation? My initial idea was to just return `(0, None)` here, since we can construct a slice of bytes from nothing

Comment (nagisa) [reply src/lib.rs:636]: a4bbf9c1 added the method/impl to `Vec` cc @fitzgen

This impl effectively means `(8, None)` is always produced. Given the documentation on the `size_hint` method, I imagine `Arbitrary for Vec` would generate a `usize` length first, and then the number of elements or something along those lines. As thus the implementation needed at least 8 bytes of unstructured input to operate.

File: src/lib.rs
Comment (nagisa) [inline src/lib.rs:631]: Consuming the entire unstructured buffer every time would mean that e.g.

```rust
#[derive(Arbitrary)]
struct Foo<'a> {
    bytes: &'a [u8],
    other_bytes: &'a [u9],
}
```

would always generate an empty slice for the 2nd field. You may want to do the `arbitrary_len` thing still here.

File: src/lib.rs
Comment (fitzgen) [reply src/lib.rs:636]: Yeah the `usize` size hint is for the vec's length.

File: src/lib.rs
Comment (fitzgen) [reply src/lib.rs:631]: Good eye @nagisa. Yes, we should still do the `arbitrary_len` thing, but also implement `arbitrary_take_rest` with this implementation that takes all the bytes.

File: src/lib.rs
Comment (frewsxcv) [reply src/lib.rs:636]: Thanks for clarifying! I think I understand the `usize` part now. And why do we use `size_hint::and(...)` here? Isn't this identical to just `<usize as Arbitrary>::size_hint(depth)`?

Comment (fitzgen) [reply src/lib.rs:636]: Yes it is. I'm not sure if I originally wrote this line or not, but in general I stylistically use `crate::size_hint::and(..)` even when its a no-op for nested fields or nested arbitrary calls just as a future proofing mechanism in case they grow in size sometime in the future or something. That doesn't really apply to this call, since it is using a hard coded value, rather than a nested `size_hint` call, so it should be safe to remove it.

Comment (frewsxcv) [reply src/lib.rs]: https://github.com/rust-fuzz/arbitrary/pull/67/commits/7d28b784114119f1b7a64f17e9c8361eb729905f

File: src/lib.rs
Comment (frewsxcv) [reply src/lib.rs]: > Consuming the entire unstructured buffer every time would mean that e.g.
> 
> ```rust
> #[derive(Arbitrary)]
> struct Foo<'a> {
>     bytes: &'a [u8],
>     other_bytes: &'a [u9],
> }
> ```
> 
> would always generate an empty slice for the 2nd field. You may want to do the `arbitrary_len` thing still here.

great catch! fixed in https://github.com/rust-fuzz/arbitrary/pull/67/commits/f951b5abbcafb8235670aa1afe426b535cdba81f

> Good eye @nagisa. Yes, we should still do the `arbitrary_len` thing, but also implement `arbitrary_take_rest` with this implementation that takes all the bytes.

the default implementation for `arbitrary_take_rest` defers to `arbitrary`:

https://github.com/rust-fuzz/arbitrary/blob/f951b5abbcafb8235670aa1afe426b535cdba81f/src/lib.rs#L173-L180

so seems like it'll already use this code. is that what you were thinking?

**Comment by frewsxcv** [comment]
@nagisa @fitzgen Thanks for your patience, this is ready for another review. Just added a couple more commits

File: src/lib.rs
Comment (fitzgen) [reply src/lib.rs]: Not quite. It should be this:

```rust
impl<'a> Arbitrary<'a> for &'a [u8] {
    // ...
    fn arbitrary_take_rest(mut u: Unstructured<'a>) -> Result<Self> {
        Ok(u.take_rest())
    }
}
```

This will always consume the rest of the `Unstructured`'s input (which is what we want in the `_take_rest` method) without getting any kind of length prefix (which we want in regular `arbitrary`, but *not* in `arbitrary_take_rest`).

Does that clear things up? Sorry this required so many back and forths to spell out

File: src/lib.rs
Comment (fitzgen) [inline src/lib.rs:1042]: Does this test still pass with the length prefix stuff? I expect it to pass when using `arbitrary_take_rest` but not with regular `arbitrary`, where I instead I would expect that we read a length (which is read from the end of the data) and then get some prefix of `[1, 2, 3]` depending on how `arbitrary_len` maps `4` into the `0..3` range.

**Review by fitzgen — COMMENTED** [review]
LGTM modulo the `arbitrary_take_rest` stuff upthread and a question about the test below.

File: src/lib.rs
Comment (frewsxcv) [reply src/lib.rs:1046]: Check the latest commit– does that match what's in your head?

**Review by fitzgen — APPROVED** [review]
Perfect! Thanks @frewsxcv :)

**Pull request merged**
